### ISSUE TITLE
Hypervisor kernel panic when ptnet VM sending big traffic with tso/gso enabled

### ISSUE THREAD

**Issue by snowcherry** [luigirizzo/netmap#393]
As mentioned in issue #387, the commands to reproduce are:
   ```
Hypervisor A:
1. qemu-system-x86_64 xenial.img -enable-kvm -smp 4 -m 4G -vga std -device ptnet-pci,netdev=data10,mac=00:AA:BB:CC:0a:06 -netdev netmap,ifname=vale1:10,id=data10,passthrough=on -vnc 0.0.0.0:1,password -monitor stdio
2. vale-ctl -a vale1:eth0 # with promisc and all offload disabled
3. In side of VM:   
netmao.ko is loaded with default ptnet_vnet_hdr=1
iperf -c 172.10.10.5 -P 2 -i 3 -t 60    # eth0 IP: 172.10.10.20 

Hypervisor B:
iperf -s  # eth0 IP: 172.10.10.5
   ```
The first feeling is the transmitting is very slow but indeed it is a kernel panic because hypervisor is dead soon. The attachment is the screenshot but unfortunately I can only get this last screen, the callstack is incomplete, just hope it can help.

![issue](https://user-images.githubusercontent.com/12002074/32474025-5c7dc6da-c3a5-11e7-9404-95098da8918a.png)

**Comment by vmaffione** [comment]
Yes this portion of stack trace unfortunately refers to something really unrelated to netmap (Linux scheduler code).
There is something missing in the description: which interface is eth0? Are you using emulated netmap or native netmap on eth0? Which commands did you use exactly to disable offloadings and enabling promisc mode?

**Comment by snowcherry** [comment]
eth0 is a physical nic in hypervisor, attached to vale1, and the vm is launched with vale1:10, so they are bridged in the same vale switch.
Command to enable promisc mode: ifconfig eth0 promisc
Command to disable offload is according to LINUX/README: ethtool -K eth0 tx off rx off gso off tso off gro off lro off

**Comment by vmaffione** [comment]
Which driver is the nic using?

**Comment by nanjj** [comment]
ixgbe 5.3.3

**Comment by vmaffione** [comment]
ok, but have you loaded the ixgbe netmap-patched driver or are you using the unmodified driver (i.e. through emulated netmap adapter)?

**Comment by nanjj** [comment]
Using the patched driver

**Comment by vmaffione** [comment]
I'm trying to reproduce your bug (using i40e NIC, I don't have ixgbe right now), but no luck.
I do see that the performance is horrible (<100Mbps), but this is a consequence of using the slowpath (virtio-net header mismatch) instead of the fastpath. The reason why there is such a slowpath is that netmap decided not to support TCP offloadings (this may change in future), because they are not needed in the use-cases that netmap addresses (middleboxes vs end user applications).
Please have a read at this document https://github.com/luigirizzo/netmap/blob/master/README.ptnetmap (specially sec. 6) to understand the motivations.
In short, using `ptnet_vnet_hdr=0` will give you decent performance (5-10Gbps).

In any case, 100Mbps or less is too slow even for a slowpath, so there is some other issue. I'll try to investigate.

**Comment by vmaffione** [comment]
Hi, I found a silly bug that was causing massive packet drops on the slowpath (VALE vnet-hdr mismatch datapath).
Now in the very same setup as yours (except for i40e NIC) I get about 23 Gbps of TCP throughput.
So in the end not so slowpath!

Could you please retry with the current master and see if you get close to line rate with the ixgbe? Thanks

**Comment by vmaffione** [comment]
Can you still reproduce this?

**Comment by vmaffione** [comment]
Please open this again if you can still reproduce it.

**Comment by snowcherry** [comment]
Thanks a lot for your effort, I'm recently writing code of vxlan termination with netmap, will try the latest version right now.

**Comment by vmaffione** [comment]
Sounds good, that's the way to go.

**Comment by snowcherry** [comment]
Yes, the problem is gone, kernel panic doesn't happen anymore, and the speed is normal, iperf 8.46Gbps.

**Comment by vmaffione** [comment]
Ok, thanks for confirming. I assume you were testing with `ptnet_vnet_hdr=1`.

**Comment by snowcherry** [comment]
[ 1111.686146] BUG: unable to handle kernel paging request at ffffc90006868000
[ 1111.686930] IP: [<ffffffff81408c42>] __memcpy+0x12/0x20
[ 1111.687459] PGD 13b099067 PUD 13b09a067 PMD ba1d4067 PTE 0
[ 1111.688061] Oops: 0000 [#1] SMP 
[ 1111.688410] Modules linked in: ppdev joydev input_leds parport_pc serio_raw parport ib_iser rdma_cm iw_cm ib_cm ib_sa ib_mad ib_core ib_addr iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsy
[ 1111.692012] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G           OE   4.4.0-96-generic #119-Ubuntu
[ 1111.692855] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.10.0-74-gcd47172a67-prebuilt.qemu-project.org 04/01/2014
[ 1111.694033] task: ffffffff81e11500 ti: ffffffff81e00000 task.ti: ffffffff81e00000
[ 1111.694760] RIP: 0010:[<ffffffff81408c42>]  [<ffffffff81408c42>] __memcpy+0x12/0x20
[ 1111.695527] RSP: 0018:ffff88013fc03dd8  EFLAGS: 00010202
[ 1111.696045] RAX: ffff8800bb7ce8c0 RBX: 00000000fffffff4 RCX: 1fffffffffb0f600
[ 1111.696739] RDX: 0000000000000004 RSI: ffffc90006867ffc RDI: ffff8800bdf538b0
[ 1111.697435] RBP: ffff88013fc03e80 R08: 0000000000000040 R09: ffff8800bb712200
[ 1111.698256] R10: 000000000000000c R11: 0000000000002000 R12: ffffc900040e300c
[ 1111.699063] R13: 0000000000000000 R14: 0000000000000000 R15: ffff8800ba98b9f0
[ 1111.699870] FS:  0000000000000000(0000) GS:ffff88013fc00000(0000) knlGS:0000000000000000
[ 1111.700876] CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
[ 1111.701547] CR2: ffffc90006868000 CR3: 00000000bb70a000 CR4: 00000000000006f0
[ 1111.702355] Stack:
[ 1111.702670]  ffffffffc01f364e ffff8800ba98b840 ffff8800b9ebeb80 ffff8800ba98c020
[ 1111.703664]  ffffc900040e3000 000000400000000c 000003ff000000f6 ffff8800ba98f000
[ 1111.704655]  ffffc90004004060 ffffc90004003000 ffff8800bb712200 00000000000000f5
[ 1111.705649] Call Trace:
[ 1111.706004]  <IRQ> 
[ 1111.706216]  [<ffffffffc01f364e>] ? ptnet_rx_poll+0x1ce/0x7a0 [netmap]
[ 1111.707105]  [<ffffffff817380fe>] net_rx_action+0x21e/0x360
[ 1111.707765]  [<ffffffff81085dc1>] __do_softirq+0x101/0x290
[ 1111.708414]  [<ffffffff810860c3>] irq_exit+0xa3/0xb0
[ 1111.709013]  [<ffffffff81845c64>] do_IRQ+0x54/0xd0
[ 1111.709597]  [<ffffffff81843d42>] common_interrupt+0x82/0x82
[ 1111.710264]  <EOI> 
[ 1111.710462]  [<ffffffff81064636>] ? native_safe_halt+0x6/0x10
[ 1111.711279]  [<ffffffff81038e5e>] default_idle+0x1e/0xe0
[ 1111.711910]  [<ffffffff8103966f>] arch_cpu_idle+0xf/0x20
[ 1111.712541]  [<ffffffff810c477a>] default_idle_call+0x2a/0x40
[ 1111.713214]  [<ffffffff810c4ae1>] cpu_startup_entry+0x2f1/0x350
[ 1111.713905]  [<ffffffff81836acc>] rest_init+0x7c/0x80
[ 1111.714520]  [<ffffffff81f5f011>] start_kernel+0x481/0x4a2
[ 1111.715170]  [<ffffffff81f5e120>] ? early_idt_handler_array+0x120/0x120
[ 1111.715927]  [<ffffffff81f5e339>] x86_64_start_reservations+0x2a/0x2c
[ 1111.716666]  [<ffffffff81f5e485>] x86_64_start_kernel+0x14a/0x16d
[ 1111.717371] Code: 75 05 e8 72 fb ff ff 48 8b 43 60 48 2b 43 50 88 43 4e 5b 5d f3 c3 90 90 90 0f 1f 44 00 00 48 89 f8 48 89 d1 48 c1 e9 03 83 e2 07 <f3> 48 a5 89 d1 f3 a4 c3 66 0f 1f 44 00 00 4 
[ 1111.720442] RIP  [<ffffffff81408c42>] __memcpy+0x12/0x20
[ 1111.721089]  RSP <ffff88013fc03dd8>
[ 1111.721545] CR2: ffffc90006868000
[ 1111.722232] ---[ end trace 652e527d311b483b ]---
[ 1111.722831] Kernel panic - not syncing: Fatal exception in interrupt
[ 1111.723976] Kernel Offset: disabled
[ 1111.724466] ---[ end Kernel panic - not syncing: Fatal exception in interrupt

**Comment by snowcherry** [comment]
Also I got this kernel panic in VM when doing receive (iperf -s)
The way to reproduce is to do iperf -c 172.100.100.12 -P 4 -i 3 -M 400 on another machine, note: this only happens with parameter -M 400, not other size!

[   19.827289] BUG: unable to handle kernel paging request at ffff8800bb5d5000
[   19.828176] IP: [<ffffffff81408c42>] __memcpy+0x12/0x20
[   19.828785] PGD 220d067 PUD 13fffe067 PMD ba107063 PTE 80000000bb5d5161
[   19.829660] Oops: 0003 [#1] SMP 
[   19.830113] Modules linked in: ppdev joydev input_leds serio_raw parport_pc parport ib_iser rdma_cm iw_cm ib_cm ib_sa ib_mad ib_core ib_addr iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsy
[   19.835040] CPU: 0 PID: 3 Comm: ksoftirqd/0 Tainted: G           OE   4.4.0-96-generic #119-Ubuntu
[   19.835949] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.10.0-74-gcd47172a67-prebuilt.qemu-project.org 04/01/2014
[   19.837162] task: ffff88013aba1980 ti: ffff88013abe0000 task.ti: ffff88013abe0000
[   19.837925] RIP: 0010:[<ffffffff81408c42>]  [<ffffffff81408c42>] __memcpy+0x12/0x20
[   19.838755] RSP: 0018:ffff88013abe3cf0  EFLAGS: 00010202
[   19.839310] RAX: ffff8800ba39d140 RBX: 00000000fffffff4 RCX: 1fffffffffdb9026
[   19.840041] RDX: 0000000000000004 RSI: ffffc900052ed6cc RDI: ffff8800bb5d5000
[   19.840777] RBP: ffff88013abe3d98 R08: 0000000000000040 R09: ffff8800bb7af600
[   19.841616] R10: 000000000000000c R11: ffff8800ba39d840 R12: ffffc900040b580c
[   19.842453] R13: 0000000000000000 R14: 0000000000000000 R15: ffff8800b972a9f0
[   19.843290] FS:  0000000000000000(0000) GS:ffff88013fc00000(0000) knlGS:0000000000000000
[   19.844329] CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
[   19.845040] CR2: ffff8800bb5d5000 CR3: 00000000ba267000 CR4: 00000000000006f0
[   19.845883] Stack:
[   19.846232]  ffffffffc01f364e ffff8800b972a840 ffff8800ba126b80 ffff8800b972f020
[   19.847396]  ffffc900040b5800 000000400000000c 000003ff0000009b ffff88003641e000
[   19.848564]  ffffc90004003ab0 ffffc90004003000 ffff8800bb7af600 903381d70000009a
[   19.849728] Call Trace:
[   19.850129]  [<ffffffffc01f364e>] ? ptnet_rx_poll+0x1ce/0x7a0 [netmap]
[   19.850919]  [<ffffffff817380fe>] net_rx_action+0x21e/0x360
[   19.851615]  [<ffffffff81085dc1>] __do_softirq+0x101/0x290
[   19.852304]  [<ffffffff810a3f90>] ? sort_range+0x30/0x30
[   19.852980]  [<ffffffff81085f78>] run_ksoftirqd+0x28/0x50
[   19.853656]  [<ffffffff810a4095>] smpboot_thread_fn+0x105/0x160
[   19.854384]  [<ffffffff810a0ce5>] kthread+0xe5/0x100
[   19.855020]  [<ffffffff810a0c00>] ? kthread_create_on_node+0x1e0/0x1e0
[   19.855810]  [<ffffffff8184360f>] ret_from_fork+0x3f/0x70
[   19.856493]  [<ffffffff810a0c00>] ? kthread_create_on_node+0x1e0/0x1e0
[   19.857283] Code: 75 05 e8 72 fb ff ff 48 8b 43 60 48 2b 43 50 88 43 4e 5b 5d f3 c3 90 90 90 0f 1f 44 00 00 48 89 f8 48 89 d1 48 c1 e9 03 83 e2 07 <f3> 48 a5 89 d1 f3 a4 c3 66 0f 1f 44 00 00 4 
[   19.862595] RIP  [<ffffffff81408c42>] __memcpy+0x12/0x20
[   19.863309]  RSP <ffff88013abe3cf0>
[   19.863798] CR2: ffff8800bb5d5000
[   19.864275] ---[ end trace 82f1f91e97cb0a20 ]---
[   19.864882] Kernel panic - not syncing: Fatal exception in interrupt
[   19.866026] Kernel Offset: disabled
[   19.866519] ---[ end Kernel panic - not syncing: Fatal exception in interrupt
[   19.867372] ------------[ cut here ]------------
[   19.867977] WARNING: CPU: 0 PID: 3 at /build/linux-z2ccW0/linux-4.4.0/arch/x86/kernel/smp.c:125 native_smp_send_reschedule+0x60/0x70()
[   19.869412] Modules linked in: ppdev joydev input_leds serio_raw parport_pc parport ib_iser rdma_cm iw_cm ib_cm ib_sa ib_mad ib_core ib_addr iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsy
[   19.874869] CPU: 0 PID: 3 Comm: ksoftirqd/0 Tainted: G      D    OE   4.4.0-96-generic #119-Ubuntu
[   19.875998] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.10.0-74-gcd47172a67-prebuilt.qemu-project.org 04/01/2014
[   19.877431]  0000000000000086 903381d7cd01c5de ffff88013fc03db0 ffffffff813fabd3
[   19.878593]  0000000000000000 ffffffff81cade78 ffff88013fc03de8 ffffffff810812e2
[   19.879755]  0000000000000001 ffff88013fc96e00 0000000000000000 ffff88013aba1980
[   19.880920] Call Trace:
[   19.881310]  <IRQ>  [<ffffffff813fabd3>] dump_stack+0x63/0x90
[   19.882068]  [<ffffffff810812e2>] warn_slowpath_common+0x82/0xc0
[   19.882802]  [<ffffffff8108142a>] warn_slowpath_null+0x1a/0x20
[   19.883520]  [<ffffffff81050a50>] native_smp_send_reschedule+0x60/0x70
[   19.884304]  [<ffffffff810bf163>] trigger_load_balance+0x133/0x210
[   19.885060]  [<ffffffff810adc16>] scheduler_tick+0xa6/0xd0
[   19.885744]  [<ffffffff810ff300>] ? tick_sched_handle.isra.14+0x60/0x60
[   19.886537]  [<ffffffff810ef5e1>] update_process_times+0x51/0x60
[   19.887272]  [<ffffffff810ff2c5>] tick_sched_handle.isra.14+0x25/0x60
[   19.888049]  [<ffffffff810ff33d>] tick_sched_timer+0x3d/0x70
[   19.888755]  [<ffffffff810eff02>] __hrtimer_run_queues+0x102/0x290
[   19.889508]  [<ffffffff810f06c8>] hrtimer_interrupt+0xa8/0x1a0
[   19.890225]  [<ffffffff81053078>] local_apic_timer_interrupt+0x38/0x60
[   19.891008]  [<ffffffff81845d1d>] smp_apic_timer_interrupt+0x3d/0x50
[   19.891774]  [<ffffffff81843fe2>] apic_timer_interrupt+0x82/0x90
[   19.892508]  <EOI>  [<ffffffff8118da69>] ? panic+0x1d5/0x215
[   19.893260]  [<ffffffff8118da62>] ? panic+0x1ce/0x215
[   19.893902]  [<ffffffff81031c6a>] oops_end+0xca/0xd0
[   19.894537]  [<ffffffff8106ad85>] no_context+0x135/0x380
[   19.895202]  [<ffffffff8106b050>] __bad_area_nosemaphore+0x80/0x1f0
[   19.895962]  [<ffffffff81724345>] ? __skb_checksum+0x65/0x2d0
[   19.896673]  [<ffffffff8106b1d3>] bad_area_nosemaphore+0x13/0x20
[   19.897408]  [<ffffffff8106b497>] __do_page_fault+0xb7/0x400
[   19.898111]  [<ffffffff8106b847>] trace_do_page_fault+0x37/0xe0
[   19.898839]  [<ffffffff81063f29>] do_async_page_fault+0x19/0x70
[   19.899564]  [<ffffffff81845428>] async_page_fault+0x28/0x30
[   19.900264]  [<ffffffff81408c42>] ? __memcpy+0x12/0x20
[   19.900929]  [<ffffffffc01f364e>] ? ptnet_rx_poll+0x1ce/0x7a0 [netmap]
[   19.901714]  [<ffffffff817380fe>] net_rx_action+0x21e/0x360
[   19.902406]  [<ffffffff81085dc1>] __do_softirq+0x101/0x290
[   19.903090]  [<ffffffff810a3f90>] ? sort_range+0x30/0x30
[   19.903754]  [<ffffffff81085f78>] run_ksoftirqd+0x28/0x50
[   19.904428]  [<ffffffff810a4095>] smpboot_thread_fn+0x105/0x160
[   19.905159]  [<ffffffff810a0ce5>] kthread+0xe5/0x100
[   19.905793]  [<ffffffff810a0c00>] ? kthread_create_on_node+0x1e0/0x1e0
[   19.906580]  [<ffffffff8184360f>] ret_from_fork+0x3f/0x70
[   19.907257]  [<ffffffff810a0c00>] ? kthread_create_on_node+0x1e0/0x1e0
[   19.908042] ---[ end trace 82f1f91e97cb0a21 ]---

**Comment by snowcherry** [comment]
Sure, will do this.

**Comment by snowcherry** [comment]
Now the traffic is actually very fast, but I found it way faster with ptnet_vnet_hdr=1 even when cross machines, what is the magic there? I remember you mentioned with ptnet_vnet_hdr=1, the offload is a slow path, but from what I am testing with my clan vxlan termination program across hypervisors, perf does really fast even with -M 100, it appears the vale port doesn't more than a tap device.

**Comment by snowcherry** [comment]
Another question is that the document says that the net map patched driver doesn't support checksum offload, but from what I am testing, it actually supports, I disable the udp checksum in my program, and I found another normal linux machine can just receive the packets without any problem.

**Comment by vmaffione** [comment]
For short packets, vale normal datapath (fastpath) typically achieves 20-30 mpps when forwarding between two ports. However, if source and destination port have different virt_hdr_len (e.g. one 0 and the other 12), you have a mismatch (slowpath),which typically halves the mpps. Note that you can have or not have mismatch both with ptnet_vnet_hdr being true or false, depending on the other vale port with which you are communicating.

If you are using large packets (e.g. tcp throughput tests) the mismatch thing has less impact on throughput, since packet rates are way lower, say 30kpps-1mpps. What matters here is TSO offloading, which allows the guest OS and vitrual switch to process 64KB packets instead of 1500 bytes packets, which means way less overhead. Now, ptnet_vnet_hdr=1 means TSO is on, hence you gbps numbers are high. Note the difference between mpps and gbps, most of confusion comes from mot realizing the difference between these two metrics, and how they depend on packet size.

How does your vxlan program interact with vale, if it does? Iow, how are you connecting together the different pieces?

If you are using vale, and there is a mismatch, vale will compute the checksum for you before passing the packet to the patched nic driver. So the patched nic driver does not do any offloading.

What is the meaning of iper M option?

**Comment by vmaffione** [comment]
In addition to TSO, there is also UFO for UDP, which is the same concept.

**Comment by snowcherry** [comment]
-M option for iperf means to specify MTU, but since the offload is on anyway, so that option doesn't seem to be really count. When I was using gdb to debug each receiving in another vale port, it actually receives 1514 bytes which means the vale port really does the segmentation offload. So this part is clear, thanks.

I understand the difference between pps and TCP stream, that's why I always test both. My program works this way, open eth0 in netmap mode, the VMs are launched with a vale switch port in ptnet bus, and my program opens another port in the same vale switch to forward the traffic, for example, vale5000 works for  vxlan 5000, vale5000:10 is a port for VM, and vxlan5000:vx5000 is used by my program to read/write traffic from/to the VMs in this vxlan. And my program also does the job of interpreting incoming vxlan packets from eth0 or packing and sending the outgoing vxlan udp packets through eth0.

So the VM traffic is added checksum when forwarded by vale, right? The other port is opened this way by my program:  
```
nmd = nm_open(netmap_port, NULL, 0, NULL);
```
I guess this means ptnet_vnet_hdr=0 because no special parameters are specified and the outgoing traffic does have the right checksum and accepted by other normal vxlan device on another hypervisor. So is it ok to specify this port with ptnet_vnet_hdr=1 because if another side is also my own program, the checksum is not needed. Although it crosses different hypervisors, but the vxlan  outer UDP packets can have checksum which ensures the packets are not malformed. And if the outer UDP packets are dropped, the inner TCP can do re-transmit anyway. 

The checksum I asked in the previous post is actually about outer vxlan UDP packets going through eth0, the document mentioned that offload is not supported, but I tested the packets sent out with checksum being 0, the other normal linux machine simply accepted and the communication went well. Also after I enabled all offload features like tso, gso, gro, the throughput increased. Then it seems the hardware offloading features like checksum or segmentation actually work on ixgbe, so I'm confused here, can you help to explain me this?

**Comment by vmaffione** [comment]
Your architecture looks good, that's the right way to add VXLAN encapsulation (or other encapsulations) to the netmap datapath with VMs.
A `virt_hdr_len` integer field is associated to each VALE port (https://github.com/luigirizzo/netmap/blob/master/sys/dev/netmap/netmap_kern.h#L826).
When you open a VALE port, `virt_hdr_len` is 0, which is the default value.
You can change the value with an ioctl, see https://github.com/luigirizzo/netmap/blob/master/apps/pkt-gen/pkt-gen.c#L1087-L1105.
If you set `ptnet_vnet_hdr==1`in the VM, QEMU will use that ioctl() to set the `virt_hdr_len` field to `12` for the associated netmap port.
So in your case it will be `vale5000:10.virt_hdr_len=12` and `vale5000:vx5000.virt_hdr_len=0`, which means that the mismatch datapath is invoked for each packet that is forwarded between two ports.
In particular, for every packet that goes from `vale5000:10` to `vale5000:vx5000`, VALE interprets the header in the packet and does all the offloadings that are necessary (TCP checksum and/or TSO, or UDP checksum and/or UFO, no other offloadings are supported, e.g. no VXLAN outer checksum support).
If you set `vale5000:vx5000.virt_hdr_len=12`, no offloadings are performed (fastpath) and your program will see the 12 byte header on each ethernet frame.

You are using the NIC patched driver, and netmap does not program the NIC (ixgbe, i40e, etc.) to do any checksum, TSO or UFO. You can look at the `ixgbe_netmap_txsync` code to check that. So the NIC should not do any checksum on its own. If it does, that's outside of the netmap control (anyway, I don't think it does, because the driver does not tell the NIC to do so).
Point here is that inner checksums are correct because they are computed by VALE (on the mismatch datapath). So ixgbe does not need to do anything.

If you enable gro/tso etc. in the VM, you will get TCP throughput increase because of the reasons I explained in my posts above.

**Comment by snowcherry** [comment]
right, I understand the inner frame was done with tso and checksum offload because of the header mismatch, but the outgoing traffic is UDP vxlan traffic packed by my program and I just put the UDP checksum as 0, another normal machine needs to accept the UDP first and then interpret the format to get inner frames, I will check more about it. And assume it doesn't, is there a way to tell a driver to do so in theory? And more over, some Nic can do vxlan offload as well, if I can enable this, probably beither inner frame nor outer frame needs to do checksum offload by software, it is still better if there is a way to leverage NIC's capability.

**Comment by vmaffione** [comment]
The stock NIC drivers are already able to do all the offloadings offered by the NIC (even VXLAN offloadings), at the cost of more code in the datapath (critical code path). We don't do it in netmap to keep the code simple and fast. (Recall that netmap patches O.S. drivers to replace the TX/RX datapath).
We may incorporate some limited offloading capabilities in future, e.g. have a look at this https://github.com/NTAP/warpcore (they extended netmap to support UDP checksum offloadings). But for now there is no support.

**Comment by snowcherry** [comment]
Thanks a lot, will look into the patch and investigate more about offloading, it does appear that the hardware helps the offloading already for UDP.

**Comment by snowcherry** [comment]
I read the vxlan spec and found vxlan UDP packet is allowed to have 0 checksum, this explains the confusion why normal linux machines accept the packets. Thank you very much.
